types/any_type.go

```diff
@@ -0,0 +1,58 @@
+package types
+
+import (
+	"fmt"
+
+	"github.com/substrait-io/substrait-go/proto"
+)
+
+// AnyType to represent AnyType, this type is to indicate "any" type of argument
+// This type is not used in function invocation. It is only used in function definition
+type AnyType struct {
+	Name        string
+	Nullability Nullability
+}
+
+func (AnyType) isRootRef() {}
+func (m AnyType) WithNullability(nullability Nullability) Type {
+	m.Nullability = nullability
+	return m
+}
+func (m AnyType) GetType() Type { return m }
+func (m AnyType) GetNullability() Nullability {
+	return m.Nullability
+}
+func (AnyType) GetTypeVariationReference() uint32 {
+	panic("not allowed")
+}
+func (AnyType) Equals(rhs Type) bool {
+	// equal to every other type
+	return true
+}
+
+func (AnyType) ToProtoFuncArg() *proto.FunctionArgument {
+	panic("not allowed")
+}
+
+func (AnyType) ToProto() *proto.Type {
+	panic("not allowed")
+}
```

Comment:
It feels like these should be removed from this type entirely and we should perhaps split the Type interface to create an interface that contains *just* the `ToProtoFuncArg` and `ToProto` methods, and then just use type assertion to do the serialization, enabling proper error reporting if the type doesn't meet the interface (rather than us just panic'ing if they use an AnyType). What do you think?

Reply:
agree, it is much cleaner this way. I was initially contemplating to do it as part of this PR or separate PR. I now fixed as part of this PR. Basically introduced interface "FuncArgType" which represents a type which can be function argument (basically can be on wire)